appinventor/components/src/com/google/appinventor/components/runtime/ListAdapterWithRecyclerView.java

```diff
@@ -1,434 +1,375 @@
-      }
-    }
-
-    @Override
-    public void onClick(View v) {
-      int position = getAdapterPosition();
-      if (multiSelect) {
-        changeSelections(position);
-      } else {
-        toggleSelection(position);
-      }
-      clickListener.onItemClick(position, v);
-    }
-  }
-
-  public void setOnItemClickListener(ClickListener clickListener) {
-    this.clickListener = clickListener;
-  }
-
-  public interface ClickListener {
-    void onItemClick(int position, View v);
-  }
-
-  public String getSelectedItems() {
-    StringBuilder sb = new StringBuilder();
-    String sep = "";
-    for (int i = 0; i < selection.length; ++i) {
-      if (selection[i]) {
-        YailDictionary dictItem = items.get(i);
-        sb.append(sep);
-        sb.append(dictItem.get(Component.LISTVIEW_KEY_MAIN_TEXT).toString());
-        sep = ",";
-      }
-    }
-    return sb.toString();
-  }
-
-  @Override
-  public Filter getFilter() {
-    return filter;
-  }
-}
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright 2024 MIT, All rights reserved
+// Released under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package com.google.appinventor.components.runtime;
+
+import android.graphics.drawable.BitmapDrawable;
+import android.graphics.drawable.Drawable;
+
+import android.util.Log;
+
+import android.view.View;
+import android.view.ViewGroup;
+import android.view.Gravity;
+
+import android.widget.Filter;
+import android.widget.Filterable;
+import android.widget.ImageView;
+import android.widget.LinearLayout;
+import android.widget.TextView;
+
+import androidx.cardview.widget.CardView;
+
+import androidx.core.view.ViewCompat;
+
+import androidx.recyclerview.widget.RecyclerView;
+
+import com.google.appinventor.components.runtime.util.MediaUtil;
+import com.google.appinventor.components.runtime.util.TextViewUtil;
+import com.google.appinventor.components.runtime.util.ViewUtil;
+import com.google.appinventor.components.runtime.util.YailDictionary;
+import com.google.appinventor.components.runtime.util.YailList;
+
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class ListAdapterWithRecyclerView
+    extends RecyclerView.Adapter<ListAdapterWithRecyclerView.RvViewHolder>  implements Filterable {
+  private static final String LOG_TAG = "ListAdapterRecyclerView";
+
+  private ClickListener clickListener;
+
+  private int textMainColor;
+  private float textMainSize;
+  private int textDetailColor;
+  private float textDetailSize;
+  private String textMainFont;
+  private String textDetailFont;
+  private int layoutType;
+  private int backgroundColor;
+  private int selectionColor;
+  private int imageHeight;
+  private int imageWidth;
+  private float radius;
+  private List<Object> items = new ArrayList<>();
+  private List<Object> oryginalItems = new ArrayList<>();
+  private List<Integer> oryginalPositions = new ArrayList<>();

```

Comment:
Yes. The reason was efficiency. In several places in the code, lists of data entered were processed in loops to check the type of list, and possibly convert one type of list to another. this was done in the ListView class and then passed to the adapter as a list of strings or a list of dictionaries. I think it took unnecessary time, which in the case of large lists stopped the user interface for even a few seconds. You wrote that you would prefer to check the type of the data list, but I found no good reason to do this and reduce performance. All data operations should take place in onBindViewHolder, which ensures efficiency even with huge data lists. Now the data in the ListView class remains unmodified. They are those entered by the user, and returned from the Elements block in the same form. A possible modification is adding or removing elements using the AddItem, RemoveItem methods, etc. Only in the adapter itself, when scrolling the list, the data list elements are checked and appropriately assigned to TextViews to display them to the user in the view.

Reply:
The reason it was written this way was that we're really supporting two kinds of behavior -- legacy, which is the comma-delimited list of strings shown in the ElementsFromString property, and current functionality, which is the multi-field YailDictionary representation. We don't want to mix-and-match these behaviors. The entire ListView should be one or the other. In the existing implementation, the check should be happening once/operation rather than once/element, which is what's happening now.

There may be (probably is) a cleaner way to approach this. ElementsFromString support is included to maintain compatibility with applications that predate the ListView overhaul, but it's not something we want to encourage for new applications.

Reply:
> There is still a question, what if Ai2 and Kodular merge? Kodular has kept the old version of ListView and added a new component that handles dictionaries. 

We're in the process of importing Kodular features into App Inventor where they make sense. If it is more logical to use the Kodular component, we should discuss that, and I can look at the code involved. This ListView was written before Kodular suggested merging, and I'm skeptical that we can separate the functionality out into two different App Inventor components without causing an upgrade headache for existing users.

> How will projects that contain one or the other Kodular List component be imported?

I don't think we expect to be able to import every Kodular app into App Inventor. One issue unrelated to the ListView is that we have no plans to ever support monetization by ads. I'm not part of the conversation about the future of Kodular aps, so I don't know what plans there are.